### ISSUE TITLE
plt.annotate() FancyArrowPatch does not respect patchA, patchB clipping

### ISSUE THREAD

**Issue by mpharrigan** [matplotlib/matplotlib#6162]
- [x] python 3.5, matplotlib 1.5.1
- [x] from miniconda (latest)

Sample code (sorry it's long, but it _is_ self contained). Please look at where I use `patchA` and `patchB` with my arrow props in annotate

``` python
from matplotlib import pyplot as plt
import itertools

patches = {}
positions = {
        'box 1': [0,0],
        'box 2': [1,0],
        'box 3': [0,1],
}

# Points
for st, (x,y) in positions.items():
    patches[st] = plt.text(x, y, st,
         ha='center', va='center',
         size= 20,
         bbox={'alpha': 0.8},
    )

# Arrows
for r, c in itertools.product(positions.keys(), positions.keys()):
    if r == c:
        continue # skip self-connections

    # Draw arrows
    x1 = positions[c]
    x2 = positions[r]
    plt.annotate('', x2, xytext=x1,
                 arrowprops=dict(
                    arrowstyle='->',
                    connectionstyle='arc3,rad=.2',
                    # THIS IS THE IMPORTANT PART vvv
                    patchA=patches[c],
                    patchB=patches[r],
                    # ^^^^^^^^^^
                 ),
                 zorder=-1,
    )

plt.xlim((-0.5,1.5))
plt.ylim(-0.5, 1.5)
plt.savefig("plot.pdf")
plt.savefig("plot.png")
```
# png

![screenshot from 2016-03-15 14 20 21](https://cloud.githubusercontent.com/assets/4967059/13794459/1dee8a6a-eab9-11e5-8b0b-27a1d297646f.png)
# pdf

![screenshot from 2016-03-15 14 20 10](https://cloud.githubusercontent.com/assets/4967059/13794458/18d16264-eab9-11e5-9af5-f4e68c5001d4.png)

I expect the arrows to be "clipped" by the boxes at the start and end of each arrow. The pdf (and svg) exported version doesn't clip at all. The png version does something weird. 

I also tried the "cairo" backend. It also performs no clipping [but changes the bend of the arrows]. Interestingly, the only thing that seems to work is the `%matplotlib inline` ipython notebook backend. It renders a png exactly as I would expect.

**Comment by tacaswell** [comment]
It is odd that %inline works, but just saving does not as it is going through the same code path.

This makes me think that this is a DPI related bug and in probably related to https://github.com/matplotlib/matplotlib/pull/6054

**Comment by tacaswell** [comment]
Interactively it looks ok, then I save it and I get the same strange arrows

![so](https://cloud.githubusercontent.com/assets/199813/13825780/9b873978-eb8a-11e5-9610-3c46441afa4f.png)

if you set the figure dpi to 100 and saving gives the correct result (but temporarily get the same funny issue in the interactive window.

![so](https://cloud.githubusercontent.com/assets/199813/13825859/021be486-eb8b-11e5-93bb-5a62b96eae61.png)


If you make sure the z-order of the arrows is above that of the patches it also seems to work (here -40 for the text, 1 for the arrows.


![so](https://cloud.githubusercontent.com/assets/199813/13826124/48f3d836-eb8c-11e5-90be-08e2c23a7f7f.png)

[so.pdf](https://github.com/matplotlib/matplotlib/files/176765/so.pdf)

**Comment by mpharrigan** [comment]
wow, thanks for looking into this. Seems like I have at least one workaround now :)

**Comment by tacaswell** [comment]
tl;dr

The problem seems to be draw order related.  The root of the problem is that we do not know how big the text will be until it is drawn, but in order for the fancy arrow back off to work (end at the edge of the bounding box, not at the center of the text) they need to know how big the text is.  The draw process is a single pass operation (unless you use `tight_layout` or `bounding_box_inches='tight'` where it gets rendered twice which is probably why `%inline` worked).  Aritsts are drawn in increasing z-order and there is no clean internal way to express 'I depend on that other artist being drawn first'.

The two workarounds are:

   1.  adjust the z-order so the text is always rendered firs
   2. make sure you call save twice :wink:

**Comment by mpharrigan** [comment]
changing the z-order fixed it. calling save twice seems to shrink the arrows *too much* (when using shrinkA and shrinkB) in pdf output
